subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,558 @@
+	CBOR_TAG_URI_TSTR =    32, /* text string	URI */
+	CBOR_TAG_BASE64URL_TSTR = 33, /* text string	base64url */
+	CBOR_TAG_BASE64_TSTR =    34, /* text string	base64 */
+	CBOR_TAG_MIME_TSTR =      36, /* text string	MIME message */
+	CBOR_TAG_CBOR =        55799, /* (any)	Self-described CBOR */
+};
+
+static int put_time(struct lwm2m_output_context *out,
+		      struct lwm2m_obj_path *path, int64_t value)
+{
+	/* CBOR time output format is unspecified but SenML CBOR uses string format.
+	 * Let's stick into the same format with plain CBOR
+	 */
+	struct tm pTime;
+
+	if (gmtime_r((time_t *)&value, &pTime) != &pTime) {
+		LOG_WRN("unable to convert from secs since Epoch to a date/time construct");
+		return 0;
+	}
+
+	/* Time construct to a string. Time in UTC, offset to local time not known */
+	char time_str[sizeof("1970-01-01T00:00:00-00:00")] = {0};
+	int time_str_len = snprintk(time_str, sizeof(time_str),
+			"%04d-%02d-%02dT%02d:%02d:%02d-00:00",
+			pTime.tm_year+1900,
+			pTime.tm_mon+1,
+			pTime.tm_mday,
+			pTime.tm_hour,
+			pTime.tm_min,
+			pTime.tm_sec);
+
+	if (time_str_len < 0 || time_str_len > sizeof(time_str)-1) {
+		LOG_WRN("unable to form a date/time string");
+		return 0;
+	}
+
+
+	cbor_state_t states[1];
+
+	new_state(states, 1, CPKT_BUF_W_REGION(out->out_cpkt), 1);
+
+	/* Are tags required? Spec V1.1 leaves this unspecified but some servers require tags */
+	bool ret = tag_encode(states, CBOR_TAG_TIME_TSTR);
+
+	if (!ret) {
+		LOG_WRN("unable to encode date/time string tag");
+		return 0;
+	}
+
+	size_t cbor_tag_sz = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+
+	out->out_cpkt->offset += cbor_tag_sz;
+
+	ret  = tstrx_encode(states,
+			&(cbor_string_type_t){.value = time_str, .len = strlen(time_str)});
+	if (!ret) {
+		LOG_WRN("unable to encode date/time string");
+		return 0;
+	}
+
+	size_t cbor_str_sz = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+
+	out->out_cpkt->offset += cbor_str_sz;
+
+	return cbor_tag_sz + cbor_str_sz;
+}
+
+static int put_s64(struct lwm2m_output_context *out,
+		      struct lwm2m_obj_path *path, int64_t value)
+{
+	cbor_state_t states[1];
+	size_t payload_len = 0;
+
+	/* TODO: remove this check once the CBOR library supports 64-bit values */
+	if (value != (value & 0xFFFFFFFF)) {
+		LOG_WRN("64-bit values are not supported");
+		return 0;
+	}
+
+
+	new_state(states, 1, CPKT_BUF_W_REGION(out->out_cpkt), 1);
+
+	/* TODO: use intx64_encode once/if available */
+	uint32_t valueu32 = (uint32_t)value;
+	bool ret = uintx32_encode(states, &valueu32);
+
+	if (ret) {
+		payload_len = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+		out->out_cpkt->offset += payload_len;
+	}
+
+	return payload_len;
+}
+
+static int put_s32(struct lwm2m_output_context *out,
+		      struct lwm2m_obj_path *path, int32_t value)
+{
+	cbor_state_t states[1];
+	size_t payload_len = 0;
+
+	new_state(states, 1, CPKT_BUF_W_REGION(out->out_cpkt), 1);
+
+	bool ret = intx32_encode(states, &value);
+
+	if (ret) {
```

Comment:
shuold we report negative if ret is false

Reply:
Fixed.

Reply:
Fixed.